### PR TITLE
[TT-Train] add assign function

### PR DESCRIPTION
**Pull request by dmakoviichuk-tt**
### Problem description
We can only create new tensor and it is impossible to update existing one in python

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes

### CODE REVIEW

File: tt-train/sources/ttml/autograd/tensor.hpp [hunk 1]
```diff
@@ -12,6 +12,8 @@
 #include "graph.hpp"
 
 namespace ttml::autograd {
+class Tensor;
```
Comment (ayerofieiev-tt) [inline tt-train/sources/ttml/autograd/tensor.hpp]: can't move using TensorPtr to after the class? 
just curious

File: tt-train/sources/ttml/autograd/tensor.hpp [hunk 2]
```diff
@@ -48,12 +50,12 @@ class Tensor : public std::enable_shared_from_this<Tensor> {
 
     bool is_grad_initialized() const;
 
+    void assign(const TensorPtr &other);
```
Comment (ayerofieiev-tt) [inline tt-train/sources/ttml/autograd/tensor.hpp]: it is a bit surprising to me that this call will assert for the shape to be the same as in currently assigned tensor.

File: tt-train/sources/ttml/autograd/tensor.hpp (hunk 1 above)
Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: I added function witch requires TensorPTr

File: tt-train/sources/ttml/autograd/tensor.hpp (hunk 2 above)
Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: Idea is simple I want to get value (ttnn::Tensor from one autograd and set it to other), it is easy to do it c++. BUt if I dont want to expose ttnn::Tensor for regular things I need to have this function.

File: tt-train/sources/ttml/autograd/tensor.hpp (hunk 1 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: ah, right. thanks

File: tt-train/sources/ttml/autograd/tensor.hpp (hunk 2 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: I understand the idea. Its just that it is very easy to assume its ok to use this function outside of that context.
Assign is very generic and really does not imply the equality of a shape. Just a note.

Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: In pytorch seems like you can do similar thing with direct access to the data but: `Directly assigning tensor.data of another size to a PyTorch tensor is generally not recommended and can lead to unexpected behavior, especially when dealing with tensors that require gradients. This is because .data provides a raw view of the underlying storage and bypasses the autograd engine.`

Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: I'd probably have this check for now.

**Review by ayerofieiev-tt — APPROVED** [review]

File: tt-train/sources/ttml/autograd/tensor.hpp (hunk 2 above)
Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: The check is useful as you know the context when this function is used. 
I am for sure hesitant asking to remove the check, because i bet the check would help to prevent bugs. 
Just wanted to raise the point that functions in public API tend to get used in all kinds of contexts, even outside of your original usecase.

**Review by rfurko-tt — APPROVED** [review]
LGTM, thanks for the change!

File: tt-train/sources/ttml/autograd/tensor.hpp (hunk 2 above)
Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/autograd/tensor.hpp]: both of you convinced me to remove it :-D

File: tt-train/configs/training_shakespeare_llama3_gpt2s_size.yaml
Comment (ayerofieiev-tt) [inline tt-train/configs/training_shakespeare_llama3_gpt2s_size.yaml:1]: Why is this here?

Comment (dmakoviichuk-tt) [reply tt-train/configs/training_shakespeare_llama3_gpt2s_size.yaml:1]: need to have a separate config for speedrun. Will be extended in a future with new features.
